### ISSUE TITLE
some err about caffe op with multiple gpu

### ISSUE THREAD

**Issue by dsqx71** [dmlc/mxnet#2972]
It work well when using caffe op with a  single gpu.

Howerer, when using caffe op with multiple gpus, I printed all gradients of  operators above caffe op and find them all zeros

**Comment by jermainewang** [comment]
@HrWangChengdu could you have a look.

Could you also provide a simple code snipplet that produce this problem?

**Comment by dsqx71** [comment]
data = mx.sym.Variable('data')
data = mx.sym.CaffeOp(data_0=data, num_weight=2,
                          prototxt='''layer{type:\"Convolution\" convolution_param { num_output: 128 pad: 1 kernel_size: 3 stride: 1 }}''')
...

model = mx.model.FeedForward( 		symbol=data ........)
model.fit( ..... , kvstore=local_allreduce_device)

**Comment by HrWangChengdu** [comment]
Could you run mxnet/example/caffe/caffe_net.py on multi-gpu? That works fine for me.

**Comment by dsqx71** [comment]
I run the example . when I use kv-store = 'local' ,it works fine.

But if kv-store = 'local_allreduce_device', train-top-accuracy don't increase.

**Comment by piiswrong** [comment]
@mli @tqchen

**Comment by piiswrong** [comment]
@HrWangChengdu This sound like some synchronization issue. Caffe probably use the default stream (Does it? I'm not sure) so you need to sync on the default stream at the end of caffe op

**Comment by HrWangChengdu** [comment]
@piiswrong thx. I'm checking on this.

**Comment by HrWangChengdu** [comment]
@piiswrong I guess this issue could be closed.
